pkgs/applications/networking/instant-messengers/discord/default.nix

```diff
@@ -86,43 +132,23 @@ let
     sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
   };
   package = if stdenv.hostPlatform.isLinux then ./linux.nix else ./darwin.nix;
-
-  packages = (
-    builtins.mapAttrs
-      (
-        _: value:
-        callPackage package (
-          value
-          // {
-            inherit src version branch;
-            meta = meta // {
-              mainProgram = value.binaryName;
-            };
-          }
-        )
-      )
-      {
-        stable = {
-          pname = "discord";
-          binaryName = "Discord";
-          desktopName = "Discord";
-        };
-        ptb = rec {
-          pname = "discord-ptb";
-          binaryName = if stdenv.hostPlatform.isLinux then "DiscordPTB" else desktopName;
-          desktopName = "Discord PTB";
-        };
-        canary = rec {
-          pname = "discord-canary";
-          binaryName = if stdenv.hostPlatform.isLinux then "DiscordCanary" else desktopName;
-          desktopName = "Discord Canary";
-        };
-        development = rec {
-          pname = "discord-development";
-          binaryName = if stdenv.hostPlatform.isLinux then "DiscordDevelopment" else desktopName;
-          desktopName = "Discord Development";
-        };
-      }
-  );
 in
-packages.${branch}
+lib.genAttrs [ "discord" "discord-ptb" "discord-canary" "discord-development" ] (
+  pname:
+  let
+    args = (variants.${stdenv.hostPlatform.system} or variants.default).${pname};
+  in
+  callPackage package (
+    args
+    // {
+      inherit pname;
+
+      meta = meta // {
+        mainProgram = args.binaryName;
+      };
+    }
```

Comment:
This should use `overrideAttrs`

Reply:
Note that these are technically args that the package is called with. At least the `pname` arg is actually used by the drv.